File: scripts/Path.py
#!/Users/subhodeep/venv/bin/python
'''class to generate path points for motion'''
import numpy as np

class Path():
    '''
    Parametric functions to generate path points. Also handles obstcale avoidance. 
    Returns. safe and optimal path point for traversal
    '''
    def __init__(self, shape, R=10.0, a=12.0, b=8.0, slope=0.0, intercept=0.0, v=1.0):
        self.shape = shape
        self.R = R
        self.a = a
        self.b = b
        self.slope = slope
        self.intercept = intercept
        self.v = v
        self.obstacles = 0

    @staticmethod
    def circle(t, R=10.0, v=1.0):
        r'''
        Generates a circle path and its kinematic properties.
        Returns: $x$, $y$, $\theta$ (heading), $\kappa$ (curvature)
        • reaches back at starting point at $t = R * 2\pi$ (62.8)
        '''
        omega = v / R
        x = R * np.cos(omega * t)
        y = R * np.sin(omega * t)
        theta = np.arctan2(y, x) + np.pi / 2
        kappa = 1.0 / R
        return x, y, theta

    @staticmethod
    def ellipse(t, a=12.0, b=8.0, v=1.0):
        r'''
        Generates an ellipse path and its kinematic properties.
        Returns: $x$, $y$, $\theta$ (heading), $\kappa$ (curvature)
        '''
        omega = v / ((a + b) / 2.0)
        x = a * np.cos(omega * t)
        y = b * np.sin(omega * t)

        dx = -a * omega * np.sin(omega * t)
        dy = b * omega * np.cos(omega * t)
        ddx = -a * omega**2 * np.cos(omega * t)
        ddy = -b * omega**2 * np.sin(omega * t)

        theta = np.arctan2(dy, dx)
        speed_sq = dx**2 + dy**2

        if speed_sq < 1e-6:
            kappa = 0
        else:
            kappa = (dx * ddy - dy * ddx) / speed_sq**(1.5)
        return x, y, theta

    @staticmethod
    def figure_eight(t, a=10.0, v=1.0):
        r'''
        Generates a figure-eight path and its kinematic properties.
        Returns: $x$, $y$, $\theta$ (heading), $\kappa$ (curvature)
        '''
        omega = v / a

        dx = a * omega * np.cos(omega * t)
        dy = a * omega * np.cos(2 * omega * t)
        ddx = -a * omega**2 * np.sin(omega * t)
        ddy = -2 * a * omega**2 * np.sin(2 * omega * t)


        x = a * np.sin(omega * t)
        y = a * np.sin(omega * t) * np.cos(omega * t)
        theta = np.arctan2(dy, dx)
        speed_sq = dx**2 + dy**2

        if speed_sq < 1e-6:
            kappa = 0
        else:
            kappa = (dx * ddy - dy * ddx) / speed_sq**(1.5)
        return x, y, theta

    @staticmethod
    def straight_line(t, slope=0.0, intercept=0.0, v=1.0):
        r'''
        Generates a straight line path and its kinematic properties.
        Returns: $x$, $y$, $\theta$ (heading), $\kappa$ (curvature)
        '''
        x = v * t
        y = slope * x + intercept
        theta = np.arctan2(slope, 1.0)
        kappa = 0.0
        return x, y, theta

    def equation(self, t):
        try:
            if self.shape == "circle":
                return Path.circle(t, R=self.R, v=self.v)
            elif self.shape == "ellipse":
                return Path.ellipse(t, a=self.a, b=self.b, v=self.v)
            elif self.shape == "figure_eight":
                return Path.figure_eight(t, a=self.a, v=self.v)
            elif self.shape == "straight_line":
                return Path.straight_line(t, slope=self.slope, intercept=self.intercept, v=self.v)
        except:
            print("Provide correct path format!")

    def add_obstacle(self, *obstacles):
        '''
        Adds obstacle
        '''
        self.obstacles = obstacles
        pass

File: src/Path.cpp
/*
    Script for parametric equations of candidate path
    Author: Subhodeep Choudhury
    Source: https://github.com/SnyprDragun/Motion-planning-Frenet-frame
*/

#include "Path.hpp"

Path::Path(string shape, float R=10.0, float a=12.0, float b=8.0, float slope=0.0, float intercept=0.0, float v=1.0){
    this->shape = shape;
    this->R = R;
    this->a = a;
    this->b = b;
    this->slope = slope;
    this->intercept = intercept;
    this->v = v;
    this->obstacles = 0;
}

Path::~Path(){}

PathPoint Path::straight_line(double t, double slope=0.0, double intercept=0.0, double v=1.0) {
    /*
    Generates a straight line path and its kinematic properties.
    Returns: x, y, theta (heading), kappa (curvature)
    */

    PathPoint path_point;

    path_point.x = v * t;
    path_point.y = slope * path_point.x + intercept;
    path_point.theta = atan2(slope, 1.0);
    path_point.kappa = 0.0;

    return path_point;
}

PathPoint Path::circle(double t, double R=10.0, double v=1.0) {
    /*
    Generates a circle path and its kinematic properties.
    Returns: x, y, theta (heading), kappa (curvature)
    */

    PathPoint path_point;

    double omega = v / R;
    path_point.x = R * cos(omega * t);
    path_point.y = R * sin(omega * t);
    path_point.theta = atan2(path_point.y, path_point.x) + M_PI / 2;
    path_point.kappa = 1.0 / R;

    return path_point;
}

PathPoint Path::ellipse(double t, double a=10.0, double b=6.0, double v=1.0) {
    /*
    Generates an ellipse path and its kinematic properties.
    Returns: x, y, theta (heading), kappa (curvature)
    */

    PathPoint path_point;

    double omega = v / ((a + b) / 2.0);

    path_point.x = a * cos(omega * t);
    path_point.y = a * sin(omega * t);

    double dx = -a * omega * sin(omega * t);
    double dy = b * omega * cos(omega * t);
    double ddx = -a * omega * omega * cos(omega * t);
    double ddy = -b * omega * omega * sin(omega * t);

    path_point.theta = atan2(dy, dx);

    double speed = dx * dx + dy * dy;
    if (speed < 1e-6) {
        path_point.kappa = 0;
    }
    else{
        path_point.kappa = (dx * ddy - dy * ddx) / pow(speed, 1.5);
    }

    return path_point;
}

PathPoint Path::figure_eight(double t, double a=10.0, double v=1.0) {
    /*
    Generates a figure-eight path and its kinematic properties.
    Returns: x, y, theta (heading), kappa (curvature)
    */

    PathPoint path_point;

    double omega = v / a;

    double dx = a * omega * cos(omega * t);
    double dy = a * omega * sin(2 * omega * t);
    double ddx = -a * omega * omega * sin(omega * t);
    double ddy = -2 * a * omega * omega * sin(2 * omega * t);

    path_point.x = a * sin(omega * t);
    path_point.y = a * sin(omega * t) * cos(omega * t);
    path_point.theta = atan2(dy, dx);

    double speed = dx * dx + dy * dy;
    if (speed < 1e-6) {
        path_point.kappa = 0;
    }
    else{
        path_point.kappa = (dx * ddy - dy * ddx) / pow(speed, 1.5);
    }

    return path_point;
}

PathPoint Path::equation(float t){
    try {
        if (this->shape == "circle") {
            return Path::circle(t, R, v);
        } else if (shape == "ellipse") {
            return Path::ellipse(t, a, b, v);
        } else if (shape == "figure_eight") {
            return Path::figure_eight(t, a, v);
        } else if (shape == "straight_line") {
            return Path::straight_line(t, slope, intercept, v);
        } else {
            throw invalid_argument("Unknown path shape!");
        }
    } catch (const exception& e) {
        cerr << "Provide correct path format! (" << e.what() << ")\n";
        return {0.0, 0.0}; // fallback
    }
}

void Path::add_obstacle(){}
